feedback/broker.go

```diff
@@ -0,0 +1,188 @@
+	Metadata         map[string]interface{} `json:"metadata"`
+	Reason           string                 `json:"reason"`
+}
+
+// Broker receives kafka messages in its InChan, unmarshal them according to the
+// platform and routes them to the correct out channel after examining their content
+type Broker struct {
+	Logger              *log.Logger
+	Config              *viper.Viper
+	InChan              *chan QueueMessage
+	pendingMessagesWG   *sync.WaitGroup
+	InvalidTokenOutChan chan *InvalidToken
+
+	run         bool
+	stopChannel chan struct{}
+}
+
+// NewBroker creates a new Broker instance
+func NewBroker(
+	logger *log.Logger, cfg *viper.Viper,
+	inChan *chan QueueMessage,
+	pendingMessagesWG *sync.WaitGroup,
+) (*Broker, error) {
+	b := &Broker{
+		Logger:            logger,
+		Config:            cfg,
+		InChan:            inChan,
+		pendingMessagesWG: pendingMessagesWG,
+		stopChannel:       make(chan struct{}),
+	}
+
+	b.configure()
+
+	return b, nil
+}
+
+func (b *Broker) loadConfigurationDefaults() {
+	b.Config.SetDefault("feedbackListeners.broker.invalidTokenChan.size", 1000)
+}
+
+func (b *Broker) configure() {
+	b.loadConfigurationDefaults()
+
+	b.InvalidTokenOutChan = make(chan *InvalidToken, b.Config.GetInt("feedbackListeners.broker.invalidTokenChan.size"))
+}
+
+// Start starts a routine to process the Broker in channel
+func (b *Broker) Start() {
+	l := b.Logger.WithField(
+		"method", "start",
+	)
+	l.Info("starting broker")
+
+	b.run = true
+	go b.processMessages()
+}
+
+// Stop stops all routines from processing the in channel and closes all output channels
+func (b *Broker) Stop() {
+	b.run = false
+	close(b.stopChannel)
+	close(b.InvalidTokenOutChan)
+}
+
+func (b *Broker) processMessages() {
+	l := b.Logger.WithField(
+		"method`", "processMessages",
+	)
+
+	for b.run == true {
+		select {
+		case msg, ok := <-*b.InChan:
+			if ok {
+				switch msg.GetPlatform() {
+				case APNSPlatform:
+					var res structs.ResponseWithMetadata
+					err := json.Unmarshal(msg.GetValue(), &res)
+					if err != nil {
+						l.WithError(err).Error(ErrAPNSUnmarshal.Error())
+					}
+					b.routeAPNSMessage(&res, msg.GetGame())
+
+				case GCMPlatform:
+					var res gcm.CCSMessage
+					err := json.Unmarshal(msg.GetValue(), &res)
+					if err != nil {
+						l.WithError(err).Error(ErrGCMUnmarshal.Error())
+					}
+					b.routeGCMMessage(&res, msg.GetGame())
+				}
+
+				b.confirmMessage()
+			}
+
+		case <-b.stopChannel:
+			break
+		}
+
+	}
+
+	l.Info("stop processing Broker's in channel")
+}
+
+func (b *Broker) routeAPNSMessage(msg *structs.ResponseWithMetadata, game string) {
+	switch msg.Reason {
+	case apns2.ReasonBadDeviceToken, apns2.ReasonUnregistered, apns2.ReasonTopicDisallowed, apns2.ReasonDeviceTokenNotForTopic:
+		tk := &InvalidToken{
+			Token:    msg.DeviceToken,
+			Game:     game,
+			Platform: APNSPlatform,
+		}
+
+		select {
+		case b.InvalidTokenOutChan <- tk:
+		default:
+			b.Logger.Error(ErrInvalidTokenChanFull.Error())
```

Comment:
What happens when the channel is full? We'll block or just log the error and move on?

Reply:
We just log the error and move on to avoid blocking the broker. Do you think that we'd rather block it?